### ISSUE TITLE
Read/configuration signals in StandardReadable don't have to be signals

### ISSUE THREAD

**Issue by callumforrester** [bluesky/ophyd-async#192]
When making a `StandardReadable` you can provide a set of signals to be read by default e.g.

```python
class Sensor(StandardReadable):
    """A demo sensor that produces a scalar value based on X and Y Movers"""

    def __init__(self, prefix: str, name="") -> None:
        # Define some signals
        self.value = epics_signal_r(float, prefix + "Value")
        self.mode = epics_signal_rw(EnergyMode, prefix + "Mode")
        # Set name and signals for read() and read_configuration()
        self.set_readable_signals(
            read=[self.value],
            config=[self.mode],
        )
        super().__init__(name=name)
```

Now the `value` signal will be read and the output included in `sensor.read()` and ditto for the `mode` signal in `sensor.read_configuration()`. `set_readable_signals` takes a list of `SignalR`s.

However, if you make sub-devices and pass them into `set_readable_signals`, it works fine, the output of `sub_device.read()` is included in `sensor.read()` as well. The only difference is that mypy complains because the subdevices are not signals. Example:

```python
class DoubleSensor(StandardReadable):
    def __init__(self, prefix: str, name: str = "", sensor_count: int = 3) -> None:
        self.sensor_a = Sensor(prefix + "SENSOR_A:")
        self.sensor_b = Sensor(prefix + "SENSOR_B:")

        # Makes read() produce the values of all sensors
        self.set_readable_signals(
            read=[self.sensor_a, self.sensor_b],  # Mypy complains here!
        )
        super().__init__(name)

# Functionally equivalent to the class above
class DoubleSensor(StandardReadable):
    def __init__(self, prefix: str, name: str = "", sensor_count: int = 3) -> None:
        self.sensor_a = Sensor(prefix + "SENSOR_A:")
        self.sensor_b = Sensor(prefix + "SENSOR_B:")

        # Makes read() produce the values of all sensors
        self.set_readable_signals(
            read=[self.sensor_a.value, self.sensor_b.value],  # Mypy is happy here, but it is extra boilerplate and requires knowledge of the sensor class; you have to know that .value is a read signal on the sensor.
        )
        super().__init__(name)
```

We do not want this behaviour to continue because allowing something potentially useful at runtime that mypy disallows will just lead to `# type: ignores` all over the place. There are two options:

1. Check the arguments of `set_readable_signals` at runtime and raise if they are not signals
2. Change the type annotation of `set_readable_signals`

From discussion with @coretl, 1. is easier. 2. is more complete but requires work. We do not currently have a suitable type for the type signature of `set_readable_signals`, it can technically take any device with `async def read(...):` but there is not protocol for that. We would have to write and maintain one. Another option would be to take a union of `SingalR` and `StandardReadable` which would get us most of the way.

We have decided not to do 2. unless concrete use cases can be identified. Does anybody have a use case for trees of nested devices that need to be auto-read?

**Comment by DominicOram** [comment]
I think the current behaviour is fine. My only issues with your example is what do to if someone adds a new signal to `Sensor` that they want to be read. My solution would instead be to have:
```python
class DoubleSensor(StandardReadable):
    def __init__(self, prefix: str, name: str = "", sensor_count: int = 3) -> None:
        self.sensor_a = Sensor(prefix + "SENSOR_A:")
        self.sensor_b = Sensor(prefix + "SENSOR_B:")

        # Makes read() produce the values of all sensors
        self.set_readable_signals(
            read=[*self.sensor_a._read_signals, *self.sensor_b._read_signals],
        )
        super().__init__(name)
```
Then it just always picks up the `read_signals` of the children and means we're not doing anything particularly magic.

**Comment by callumforrester** [comment]
@DominicOram I think that's a good solution actually, although I'd want `read_signals` to become a public property

**Comment by DominicOram** [comment]
> @DominicOram I think that's a good solution actually, although I'd want read_signals to become a public property

Credit to @dperl-dls actually. Yes, I agree it should become public for this. We would probably want the equivalent for `config` too?

**Comment by coretl** [comment]
@callumforrester sold me on the idea of supporting `StandardDetector` as well as `SignalR` here, and now it looks like we're not doing that?

**Comment by coretl** [comment]
We should also consider how we would do #44 here. Especially how we mark signals to be read as hinted...

**Comment by DiamondJoseph** [comment]
I thought parent.read() already called child.read()?

**Comment by DiamondJoseph** [comment]
Worth considering also PVI case then, may need to re-calculate readable_signals after connect

**Comment by coretl** [comment]
Ok, after discussion with @callumforrester I think the usecase is compound devices like a set of slits. We want a way to opt into the behaviour that we got with ophyd where we aggregate all the signals our children have chosen to read. With that in mind I propose that we make `self.set_readable` take things that we can call `read`, `describe`, `read_configuration`, `describe_configuration` and `hints` on, and provide some utility functions like `config_signals` or `uncached_signals` or `hinted_signals` to wrap with the right verbs:
```python
class Motor(StandardReadable):
    def __init__(self, prefix, name=None):
        self.setpoint = epics_signal_rw(prefix + ".VAL")
        self.readback = epics_signal_rw(prefix + ".RBV")
        self.units = epics_signal_rw(prefix + ".EGU")
        self.velocity = epics_signal_rw(prefix + ".VELO")
        self.set_readable(
            *config_signals(self.units, self.velocity), 
            *hinted_signals(self.readback)
        )
        super().__init__(name)

class Slits(StandardReadable):
    def __init__(self, prefix, name=None):
        self.x_gap = Motor(prefix + ":X:GAP")
        self.y_gap = Motor(prefix + ":Y:GAP")
        self.x_centre = Motor(prefix + ":X:CENTRE")
        self.y_centre = Motor(prefix + ":Y:CENTRE")
        self.set_readable(*self.children())
        super().__init__(name)
```

**Comment by coretl** [comment]
Could even use the power of the walrus...
```python
class Motor(StandardReadable):
    def __init__(self, prefix, name=None):
        self.setpoint = epics_signal_rw(prefix + ".VAL")
        self.set_readable(
            *config_signals(
                self.units := epics_signal_rw(prefix + ".EGU"),
                self.velocity := epics_signal_rw(prefix + ".VELO"),
            ),
            *hinted_signals(
                self.readback := epics_signal_rw(prefix + ".RBV"),
            ),
        )
        super().__init__(name)

class Slits(StandardReadable):
    def __init__(self, prefix, name=None):
        self.set_readable(
            self.x_gap := Motor(prefix + ":X:GAP"),
            self.y_gap := Motor(prefix + ":Y:GAP"),
            self.x_centre := Motor(prefix + ":X:CENTRE"),
            self.y_centre := Motor(prefix + ":Y:CENTRE"),
        )
        super().__init__(name)
```

**Comment by callumforrester** [comment]
For now, I think we will just make the read signals public: https://github.com/bluesky/ophyd-async/issues/196

**Comment by coretl** [comment]
Last idea before I forget:

```python
class Motor(StandardReadable):
    def __init__(self, prefix, name=None):
        self.setpoint = epics_signal_rw(prefix + ".VAL")
        with self.add_readables(ConfigSignal):
            self.units = epics_signal_rw(prefix + ".EGU")
            self.velocity = epics_signal_rw(prefix + ".VELO")
        with self.add_readables(HintedSignal):
            self.readback = epics_signal_rw(prefix + ".RBV")
        super().__init__(name)

class Slits(StandardReadable):
    def __init__(self, prefix, name=None):
        with self.add_readables():
            self.x_gap = Motor(prefix + ":X:GAP")
            self.y_gap = Motor(prefix + ":Y:GAP")
            self.x_centre = Motor(prefix + ":X:CENTRE")
            self.y_centre = Motor(prefix + ":Y:CENTRE")
        super().__init__(name)
```

**Comment by coretl** [comment]
Show of hands, should we:
1. :tada: use `with self.add_readables(SomeWrapperClass):` as per https://github.com/bluesky/ophyd-async/issues/192#issuecomment-2051090558
2. :eyes: use `self.set_readable(*some_wrapper_function)` as per https://github.com/bluesky/ophyd-async/issues/192#issuecomment-2049796039
3. :rocket: use `self.set_readable_signals(read=..., config=...)` with public `read_signals` and `config_signals` as per https://github.com/bluesky/ophyd-async/issues/192#issue-2237360654
Please vote with the emojis below

**Comment by DominicOram** [comment]
I think 🚀  is the simplest thing to do to solve the original issue of "I want to easily pass Devices to set readables". I'm not against 🎉 but it seems to be a complete re-write of how we set signals that goes beyond the original issue

**Comment by jwlodek** [comment]
Would it be feasible to add a method that returns lists of readable/config signals in a device that can then be used instead of the device itself? That way this functionality can still fairly easily exist without needing significant refactoring to satisfy mypy.

For the record I can think of cases where something like this would be desirable. For example a device with 4 duplicate channels that offer several data collecting signals each. I'd make a class that represents each channel and the overall device would instantiate 4 - and then I'd like to be able to easily get all relevant signals.

**Comment by coretl** [comment]
The only problem with :rocket: is that we still need to find a solution that satisfies #44 . This means we will either need to write a `def hints()` function that pulls together the hints of the child functions, or change the underlying datastructures to be dictionaries instead. For completeness here are the 2 examples:
## :rocket:
```python
class Motor(StandardReadable):
    def __init__(self, prefix, name=None):
        self.setpoint = epics_signal_rw(prefix + ".VAL")
        self.units = epics_signal_rw(prefix + ".EGU")
        self.velocity = epics_signal_rw(prefix + ".VELO")
        self.readback = epics_signal_rw(prefix + ".RBV")
        self.set_readable_signals(
            read=[self.readback],
            config=[self.units, self.velocity],
        )
        super().__init__(name)

    @property
    def hints(self):
        return {"fields": [self.readback.name]}

class Slits(StandardReadable):
    def __init__(self, prefix, name=None):
        self.x_gap = Motor(prefix + ":X:GAP")
        self.y_gap = Motor(prefix + ":Y:GAP")
        self.x_centre = Motor(prefix + ":X:CENTRE")
        self.y_centre = Motor(prefix + ":Y:CENTRE")
        self.set_readable_signals(
            read=[*self.x_gap.read_signals, *self.y_gap.read_signals, *self.x_centre.read_signals, *self.y_centre.read_signals],
            config=[*self.x_gap.config_signals, *self.y_gap.config_signals, *self.x_centre.config_signals, *self.y_centre.config_signals],
        )
        super().__init__(name)

    @property
    def hints(self):
        return {**self.x_gap.hints, **self.y_gap.hints, **self.x_centre.hints, **self.y_centre.hints}
```

## :tada:
```python
class Motor(StandardReadable):
    def __init__(self, prefix, name=None):
        self.setpoint = epics_signal_rw(prefix + ".VAL")
        with self.add_readables(ConfigSignal):
            self.units = epics_signal_rw(prefix + ".EGU")
            self.velocity = epics_signal_rw(prefix + ".VELO")
        with self.add_readables(HintedSignal):
            self.readback = epics_signal_rw(prefix + ".RBV")
        super().__init__(name)

class Slits(StandardReadable):
    def __init__(self, prefix, name=None):
        with self.add_readables():
            self.x_gap = Motor(prefix + ":X:GAP")
            self.y_gap = Motor(prefix + ":Y:GAP")
            self.x_centre = Motor(prefix + ":X:CENTRE")
            self.y_centre = Motor(prefix + ":Y:CENTRE")
        super().__init__(name)
```
---
Personally I think :tada: is worth the extra effort now, if it satisfies all our use cases. The edge conditions of "I want to pick and choose between signals of child devices rather than reading the whole device" could be added with some additional functions on the `self.add_readables()` context manager...

Thoughts?

**Comment by DominicOram** [comment]
I'm happy with :tada: being done but maybe for clarity of what we're doing we could make a new issue that specifically says what we're doing and why as it's getting tangential to this?

**Comment by coretl** [comment]
Closed in favour of #223
